engine/table/src/main/java/io/deephaven/engine/table/impl/multijoin/typed/incopen/gen/IncrementalMultiJoinHasherChar.java

```diff
@@ -0,0 +1,345 @@
+        }
+    }
+
+    private static int hash(char k0) {
+        int hash = CharChunkHasher.hashInitialSingle(k0);
+        return hash;
+    }
+
+    private boolean migrateOneLocation(int locationToMigrate,
+            MultiJoinModifiedSlotTracker modifiedSlotTracker) {
+        final int currentStateValue = alternateSlotToOutputRow.getUnsafe(locationToMigrate);
+        if (currentStateValue == EMPTY_OUTPUT_ROW) {
+            return false;
+        }
+        final char k0 = alternateKeySource0.getUnsafe(locationToMigrate);
+        final int hash = hash(k0);
+        int destinationTableLocation = hashToTableLocation(hash);
+        while (slotToOutputRow.getUnsafe(destinationTableLocation) != EMPTY_OUTPUT_ROW) {
+            destinationTableLocation = nextTableLocation(destinationTableLocation);
+        }
+        mainKeySource0.set(destinationTableLocation, k0);
+        slotToOutputRow.set(destinationTableLocation, currentStateValue);
+        final int cookie  = alternateModifiedTrackerCookieSource.getUnsafe(locationToMigrate);
+        mainModifiedTrackerCookieSource.set(destinationTableLocation, cookie);
+        alternateModifiedTrackerCookieSource.set(locationToMigrate, EMPTY_COOKIE_SLOT);
+        alternateSlotToOutputRow.set(locationToMigrate, EMPTY_OUTPUT_ROW);
+        return true;
+    }
+
+    @Override
+    protected int rehashInternalPartial(int entriesToRehash,
+            MultiJoinModifiedSlotTracker modifiedSlotTracker) {
+        int rehashedEntries = 0;
+        while (rehashPointer > 0 && rehashedEntries < entriesToRehash) {
+            if (migrateOneLocation(--rehashPointer, modifiedSlotTracker)) {
+                rehashedEntries++;
+            }
+        }
+        return rehashedEntries;
+    }
+
+    @Override
+    protected void newAlternate() {
+        super.newAlternate();
+        this.mainKeySource0 = (ImmutableCharArraySource)super.mainKeySources[0];
+        this.alternateKeySource0 = (ImmutableCharArraySource)super.alternateKeySources[0];
+    }
+
+    @Override
+    protected void clearAlternate() {
+        super.clearAlternate();
+        this.alternateKeySource0 = null;
+    }
+
+    @Override
+    protected void migrateFront(MultiJoinModifiedSlotTracker modifiedSlotTracker) {
+        int location = 0;
+        while (migrateOneLocation(location++, modifiedSlotTracker));
+    }
+
+    @Override
+    protected void rehashInternalFull(final int oldSize) {
+        final char[] destKeyArray0 = new char[tableSize];
+        final int[] destState = new int[tableSize];
+        Arrays.fill(destState, EMPTY_OUTPUT_ROW);
+        final char [] originalKeyArray0 = mainKeySource0.getArray();
+        mainKeySource0.setArray(destKeyArray0);
+        final int [] originalStateArray = slotToOutputRow.getArray();
+        slotToOutputRow.setArray(destState);
+        final int [] oldModifiedCookie = mainModifiedTrackerCookieSource.getArray();
+        final int [] destModifiedCookie = new int[tableSize];
```

Comment:
I think this suggests cookie generation should be initialized to 1, at construction or at wraparound.